cpp/tests/mtmg/threaded_test.cpp

```diff
@@ -0,0 +1,469 @@
+          thread_handle.raft_handle(),
+          graph_view.get_pointer(thread_handle).value(),
+          std::nullopt,
+          std::nullopt,
+          std::nullopt,
+          std::nullopt,
+          std::nullopt,
+          pageranks.get_pointer(thread_handle).value(),
+          alpha,
+          epsilon,
+          std::numeric_limits<size_t>::max(),
+          false,
+          false);
+#endif
+      });
+    }
+
+    // Wait for CPU threads to complete
+    std::for_each(running_threads.begin(), running_threads.end(), [](auto& t) { t.join(); });
+    running_threads.resize(0);
+    instance_manager->reset_threads();
+
+    std::vector<std::tuple<std::vector<vertex_t>, std::vector<result_t>>> computed_pageranks_v;
+    std::mutex computed_pageranks_lock{};
+
+    // Load computed_pageranks from different threads.  We'll use more threads than GPUs here
+    for (int i = 0; i < num_threads; ++i) {
+      running_threads.emplace_back([&instance_manager,
+                                    &graph_view,
+                                    &pageranks,
+                                    &computed_pageranks_lock,
+                                    &computed_pageranks_v,
+                                    &h_src_v,
+                                    &h_dst_v,
+                                    &h_weights_v,
+                                    i,
+                                    num_threads]() {
+        auto thread_handle = instance_manager->get_handle();
+
+        auto number_of_vertices = graph_view.get_pointer(thread_handle)->number_of_vertices();
+
+        std::vector<vertex_t> my_vertex_list;
+        my_vertex_list.reserve((number_of_vertices + num_threads - 1) / num_threads);
+
+        for (int j = i; j < number_of_vertices; j += num_threads) {
+          my_vertex_list.push_back(j);
+        }
+
+        rmm::device_uvector<vertex_t> d_my_vertex_list(my_vertex_list.size(),
+                                                       thread_handle.raft_handle().get_stream());
+        raft::update_device(d_my_vertex_list.data(),
+                            my_vertex_list.data(),
+                            my_vertex_list.size(),
+                            thread_handle.raft_handle().get_stream());
+
+        auto d_my_pageranks = pageranks.gather(
+          thread_handle,
+          raft::device_span<vertex_t const>{d_my_vertex_list.data(), d_my_vertex_list.size()},
+          graph_view);
+
+        std::vector<result_t> my_pageranks(d_my_pageranks.size());
+        raft::update_host(my_pageranks.data(),
+                          d_my_pageranks.data(),
+                          d_my_pageranks.size(),
+                          thread_handle.raft_handle().get_stream());
+
+        {
+          std::lock_guard<std::mutex> lock(computed_pageranks_lock);
+          computed_pageranks_v.push_back(
+            std::make_tuple(std::move(my_vertex_list), std::move(my_pageranks)));
+        }
```

Comment:
The `gather` call above will handle these in the multi-node case.

When we create the instance manager, we will configure a NCCL communications object that allows us to do point-to-point messaging between the participating GPUs (entirely in device memory).  The `gather` call above will perform the following steps:
* Shuffle the vertex ids from `d_my_vertex_list` (in device memory) to the GPU where the data for that information resides
* On each GPU, do a local gather of the vertices on that GPU
* Shuffle the vertex ids/pagerank pairs back to the original GPU

This will all be buried in that call, no extra work is required by the caller.